File: day24/main.py
from typing import List
from itertools import permutations


class ALU:
    def __init__(self, _input: List[int]) -> None:
        self.w = 0
        self.x = 0
        self.y = 0
        self.z = 0
        self.input = _input
        self.pos = 0

    def run(self, file_name):
        with open(file_name) as f:
            for line in f.readlines():
                instruction, *res = line.strip().split(" ")
                if len(res) == 1:
                    n = self.input[self.pos]
                    self.pos += 1
                    if res[0] == "w":
                        self.w = n
                    elif res[0] == "x":
                        self.x = n
                    elif res[0] == "y":
                        self.y = n
                    elif res[0] == "z":
                        self.z = n
                else:
                    if res[0] == "w":
                        a = self.w
                    elif res[0] == "x":
                        a = self.x
                    elif res[0] == "y":
                        a = self.y
                    elif res[0] == "z":
                        a = self.z

                    try:
                        b = int(res[1])
                    except:
                        if res[1] == "w":
                            b = self.w
                        elif res[1] == "x":
                            b = self.x
                        elif res[1] == "y":
                            b = self.y
                        elif res[1] == "z":
                            b = self.z

                    if instruction == "add":
                        a = a + b
                    elif instruction == "mul":
                        a = a * b
                    elif instruction == "div":
                        if b == 0:
                            print("AJJAWKEJFEKJF")
                        a = a // b
                    elif instruction == "mod":
                        if a < 0 or b <= 0:
                            print("EJJEEJEJJEEJJEJE")
                        a = a % b
                    elif instruction == "eql":
                        a = 1 if a == b else 0

                    if res[0] == "w":
                        self.w = a
                    elif res[0] == "x":
                        self.x = a
                    elif res[0] == "y":
                        self.y = a
                    elif res[0] == "z":
                        self.z = a

        return self.z

    def __str__(self) -> str:
        return f"{self.z=}\n{self.y=}\n{self.x=}\n{self.w=}\n{self.pos=}\n{self.input=}"


alu = ALU([int(k) for k in str(54986732675439)])
alu.run("./input.txt")
print(alu)

File: day24/src/main.rs
use std::cell::Cell;
use std::fs::read_to_string;

fn main() {
    let mut alu = ALU::new(vec![1, 3, 4, 5, 6, 7, 8, 9, 1, 2, 3, 4, 5, 3]);
    let mut a = &alu.pos;
    *a = 4;
    println!("{}", alu.pos);
    // 0


    // alu.run("./test.txt");
}

struct ALU {
    w: Cell<i32>,
    x: Cell<i32>,
    y: Cell<i32>,
    z: Cell<i32>,
    pos: usize,
    input: Vec<i32>,
}

impl ALU {
    fn new(input: Vec<i32>) -> Self {
        ALU {
            w: 0,
            x: 0,
            y: 0,
            z: 0,
            pos: 0,
            input,
        }
    }

    fn run(&mut self, file_name: &str) {
        for line in read_to_string(file_name).unwrap().lines() {
            let split = line.split(" ").collect::<Vec<_>>();
            let res: Option<(&str, i32, i32)>;
            if split.len() == 2 {
                match split[1] {
                    "x" => self.x = self.input[self.pos],
                    "y" => self.y = self.input[self.pos],
                    "z" => self.z = self.input[self.pos],
                    "w" => self.w = self.input[self.pos],
                    _ => unreachable!(),
                };
            } else {
                let instruction = split[0];
                let mut a = match split[1] {
                    "x" => &self.x,
                    "y" => &self.y,
                    "z" => &self.z,
                    _ => unreachable!(),
                };
                let b;
                if let Ok(x) = split[2].parse::<i32>() {
                    b = x;
                } else {
                    b = match split[2] {
                        "x" => self.x,
                        "y" => self.y,
                        "z" => self.z,
                        _ => unreachable!(),
                    };
                }

                match instruction {
                    "add" => a = &(*a + b),
                    "mul" => a = &(*a * b),
                    "div" => a = &(*a / b),
                    "mod" => a = &(*a % b),
                    "eql" => a = &(if *a == b { 1 } else { 0 }),
                    _ => unimplemented!(),
                }
            }
        }
        println!(
            "{:?} {} {} {} {} {}",
            self.input, self.w, self.x, self.y, self.z, self.pos
        );
    }
}
